scanpipe/tests/test_models.py

```diff
@@ -625,6 +627,26 @@ def test_scanpipe_codebase_resource_children(self):
         ]
         self.assertEqual(expected, [resource.path for resource in children])
 
+    def test_scanpipe_codebase_resource_walk(self):
+        asgiref_root = self.project_asgiref.codebaseresources.get(path="codebase")
+        asgiref_scan = self.data_location / "asgiref-3.3.0_scan.json"
+        asgiref_vc = VirtualCodebase(location=asgiref_scan)
+        asgiref_vc_root = asgiref_vc.root
+
+        topdown_paths = list(r.path for r in asgiref_root.walk(topdown=True))
+        # We walk from the root Resource of `asgiref_vc` to make it comparable
+        # to how we are walking the children of `asgiref_root`
+        expected_topdown_paths = list(
+            r.path for r in asgiref_vc_root.walk(asgiref_vc, topdown=True)
+        )
+        self.assertEqual(expected_topdown_paths, topdown_paths)
```

Comment:
@tdruez 

I've modified `CodebaseResource.children()` to order the children by lowercased path by adding `.order_by(Lower("path"))` to the queryset. This is done to mimic the behavior of `commoncode.resource.Resource.children()`. (https://github.com/JonoYang/scancode.io/blob/codebaseresource-walk/scanpipe/models.py#L918) However, I haven't gotten it exactly right since this test fails because the `CodebaseResource.walk()` method is visiting `codebase/asgiref-3.3.0-py3-none-any.whl-extract/asgiref/compatibility.py` after visiting `codebase/asgiref-3.3.0-py3-none-any.whl-extract/asgiref/`, whereas `commoncode.resource.Codebase.walk()` visits `codebase/asgiref-3.3.0-py3-none-any.whl-extract/asgiref/__init__.py` after visiting `codebase/asgiref-3.3.0-py3-none-any.whl-extract/asgiref/`.

```
# Paths generated from CodebaseResource.walk()
topdown_paths = [
    'codebase/asgiref-3.3.0-py3-none-any.whl',
    'codebase/asgiref-3.3.0-py3-none-any.whl-extract',
    'codebase/asgiref-3.3.0-py3-none-any.whl-extract/asgiref',
    'codebase/asgiref-3.3.0-py3-none-any.whl-extract/asgiref/compatibility.py',
    'codebase/asgiref-3.3.0-py3-none-any.whl-extract/asgiref/current_thread_executor.py',
    'codebase/asgiref-3.3.0-py3-none-any.whl-extract/asgiref/__init__.py',
    'codebase/asgiref-3.3.0-py3-none-any.whl-extract/asgiref/local.py',
    'codebase/asgiref-3.3.0-py3-none-any.whl-extract/asgiref/server.py',
    'codebase/asgiref-3.3.0-py3-none-any.whl-extract/asgiref/sync.py',
    'codebase/asgiref-3.3.0-py3-none-any.whl-extract/asgiref/testing.py',
    'codebase/asgiref-3.3.0-py3-none-any.whl-extract/asgiref/timeout.py',
    'codebase/asgiref-3.3.0-py3-none-any.whl-extract/asgiref/wsgi.py',
    'codebase/asgiref-3.3.0-py3-none-any.whl-extract/asgiref-3.3.0.dist-info',
    'codebase/asgiref-3.3.0-py3-none-any.whl-extract/asgiref-3.3.0.dist-info/LICENSE',
    'codebase/asgiref-3.3.0-py3-none-any.whl-extract/asgiref-3.3.0.dist-info/METADATA',
    'codebase/asgiref-3.3.0-py3-none-any.whl-extract/asgiref-3.3.0.dist-info/RECORD',
    'codebase/asgiref-3.3.0-py3-none-any.whl-extract/asgiref-3.3.0.dist-info/top_level.txt',
    'codebase/asgiref-3.3.0-py3-none-any.whl-extract/asgiref-3.3.0.dist-info/WHEEL'
]

# Paths generated by commoncode.resource.Codebase.walk()
expected_topdown_paths = [
    'codebase/asgiref-3.3.0-py3-none-any.whl',
    'codebase/asgiref-3.3.0-py3-none-any.whl-extract',
    'codebase/asgiref-3.3.0-py3-none-any.whl-extract/asgiref',
    'codebase/asgiref-3.3.0-py3-none-any.whl-extract/asgiref/__init__.py',
    'codebase/asgiref-3.3.0-py3-none-any.whl-extract/asgiref/compatibility.py',
    'codebase/asgiref-3.3.0-py3-none-any.whl-extract/asgiref/current_thread_executor.py',
    'codebase/asgiref-3.3.0-py3-none-any.whl-extract/asgiref/local.py',
    'codebase/asgiref-3.3.0-py3-none-any.whl-extract/asgiref/server.py',
    'codebase/asgiref-3.3.0-py3-none-any.whl-extract/asgiref/sync.py',
    'codebase/asgiref-3.3.0-py3-none-any.whl-extract/asgiref/testing.py',
    'codebase/asgiref-3.3.0-py3-none-any.whl-extract/asgiref/timeout.py',
    'codebase/asgiref-3.3.0-py3-none-any.whl-extract/asgiref/wsgi.py',
    'codebase/asgiref-3.3.0-py3-none-any.whl-extract/asgiref-3.3.0.dist-info',
    'codebase/asgiref-3.3.0-py3-none-any.whl-extract/asgiref-3.3.0.dist-info/LICENSE',
    'codebase/asgiref-3.3.0-py3-none-any.whl-extract/asgiref-3.3.0.dist-info/METADATA',
    'codebase/asgiref-3.3.0-py3-none-any.whl-extract/asgiref-3.3.0.dist-info/RECORD',
    'codebase/asgiref-3.3.0-py3-none-any.whl-extract/asgiref-3.3.0.dist-info/top_level.txt',
    'codebase/asgiref-3.3.0-py3-none-any.whl-extract/asgiref-3.3.0.dist-info/WHEEL'
]
```

Is it possible to have `.order_by()` consider non-alphabetical characters when sorting, such that `__init__.py` would be returned before `compatibility.py`?

Reply:
After talking to @pombredanne, it may be a lost cause to ensure the walking order is the same between `commoncode.resource.Resource.walk()` and `CodebaseResource.walk()`.